types/src/proof/definition.rs

```diff
@@ -470,6 +471,143 @@ impl From<AccumulatorConsistencyProof> for crate::proto::types::AccumulatorConsi
     }
 }
 
+/// A proof that is similar to `AccumulatorProof`, but can be used to authenticate a range of
+/// leaves. For example, given the following accumulator:
+///
+/// ```text
+///                 root
+///                /     \
+///              /         \
+///            /             \
+///           o               o
+///         /   \           /   \
+///        /     \         /     \
+///       X       o       o       Y
+///      / \     / \     / \     / \
+///     o   o   a   b   c   Z   o   o
+/// ```
+///
+/// if the proof wants to show that `[a, b, c]` exists in the accumulator, it would need `X` on the
+/// left and `Y` and `Z` on the right.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct AccumulatorRangeProof<H> {
+    /// The siblings on the left of the path from root to the first leaf. Siblings near the root
+    /// are at the beginning of the vector.
+    left_siblings: Vec<HashValue>,
+
+    /// The sliblings on the right of the path from root to the last leaf. Siblings near the root
+    /// are at the beginning of the vector.
+    right_siblings: Vec<HashValue>,
+
+    phantom: PhantomData<H>,
+}
+
+impl<H> AccumulatorRangeProof<H>
+where
+    H: CryptoHasher,
+{
+    /// Constructs a new `AccumulatorRangeProof` using `left_siblings` and `right_siblings`.
+    pub fn new(left_siblings: Vec<HashValue>, right_siblings: Vec<HashValue>) -> Self {
+        Self {
+            left_siblings,
+            right_siblings,
+            phantom: PhantomData,
+        }
+    }
+
+    /// Verifies the proof is correct. The verifier needs to have `expected_root_hash`, the index
+    /// of the first leaf and all of the leaves in possession.
+    pub fn verify(
+        &self,
+        expected_root_hash: HashValue,
+        first_leaf_index: Option<u64>,
+        leaf_hashes: &[HashValue],
+    ) -> Result<()> {
+        if first_leaf_index.is_none() {
+            ensure!(
+                leaf_hashes.is_empty(),
+                "first_leaf_index indicated empty list while leaf_hashes is not empty.",
+            );
+            ensure!(
+                self.left_siblings.is_empty() && self.right_siblings.is_empty(),
+                "No siblings are needed.",
+            );
+            return Ok(());
+        }
+
+        ensure!(
+            self.left_siblings.len() <= MAX_ACCUMULATOR_PROOF_DEPTH,
+            "Proof has more than {} ({}) left siblings.",
+            MAX_ACCUMULATOR_PROOF_DEPTH,
+            self.left_siblings.len(),
+        );
+        ensure!(
+            self.right_siblings.len() <= MAX_ACCUMULATOR_PROOF_DEPTH,
+            "Proof has more than {} ({}) right siblings.",
+            MAX_ACCUMULATOR_PROOF_DEPTH,
+            self.right_siblings.len(),
+        );
+        ensure!(
+            !leaf_hashes.is_empty(),
+            "leaf_hashes is empty while first_leaf_index indicated non-empty list.",
+        );
+
+        let mut left_sibling_iter = self.left_siblings.iter().rev().peekable();
```

Comment:
unrelated: we `reverse()`ed in `get_siblings()` and `rev()`s here -- seems a waste. Shall we change the proof format (separately)?

Reply:
I don't remember why we set up things in this way. @lightmark ?